File: knapsack.c
#include<stdio.h>
#include<stdlib.h>

struct kp
{
    float w;
    float p;
    float pw;
};

int main()
{
    int cap, n;
    printf("Enter capacity: ");
    scanf("%d",&cap);
    printf("No. of items: ");
    scanf("%d",&n);
    struct kp item[n];
    for(int i=0; i<n; i++)
    {
        printf("Enter weight for item %d:",i+1);
        scanf("%f",&item[i].w);
        printf("Enter profit for item %d:",i+1);
        scanf("%f",&item[i].p);
        item[i].pw=item[i].p / item[i].w;
        printf("Profit per weigth : %f\n",item[i].pw);
    }
    //sorting in desc order wrt p/w
    for(int i=0; i<n; i++)
    {
        for(int j=0; j<n-i-1; j++)
        {
            if(item[j].pw < item[j+1].pw) 
            {
                struct kp temp = item[j];
                item[j] = item[j+1];
                item[j+1] = temp;
            }
        }
    }
    int currW = 0;
    float currP = 0;
    for(int i=0; i<n; i++)
    {
        if(currW + item[i].w <= cap)
        {
            // printf("%f",item[i].pw);
            currW += item[i].w;
            currP += item[i].p;
            // printf("%f\n",currP);
        }
        else
        {
            int rem = cap - currW;
            currP += item[i].p * (rem/item[i].w);
            
            break;
        }
    }
    printf("The maximum profit is : %f",currP);
    return 0;
}

File: knapsack.java
import java.util.Scanner;

public class knapsack {
    static int MAX = 1000;
    static int max(int a, int b)
    {
        return a<b ? b:a;
    }
    static int knapsack(int p[],int w[], int cap, int n)
    {
        int profit[][] = new int [MAX][MAX];
        for(int i=0; i<=n; i++)
        {
            for(int j=0; j<=cap; j++)
            {
                if( i == 0 || j == 0 )
                {
                    profit[i][j] = 0;
                }
                else if(w[i] <= j)
                {
                    profit[i][j] = max(profit[i-1][j], profit[i-1][j-w[i]]+p[i]);
                }
                else
                {
                    profit[i][j] = profit[i-1][j];
                }
            }
        }
        return profit[n][cap];
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("No. of items: ");
        int n = sc.nextInt();
        int p[]=new int[n+1];
        int w[] = new int[n+1];
        for(int i=0; i<n; i++)
            p[i] = sc.nextInt();
        for(int i=0; i<n; i++)
            w[i] = sc.nextInt();
        System.out.println("Capacity : ");
        int cap = sc.nextInt();
        int profit = knapsack(p, w, cap, n);
        System.out.println(profit);
        sc.close();
    }
}
